Library/Homebrew/test/utils_spec.rb

```diff
@@ -0,0 +1,340 @@
+        expect(dir/"called").to exist
+      end
+    end
+
+    specify "#with_custom_locale" do
+      ENV["LC_ALL"] = "en_US.UTF-8"
+
+      with_custom_locale("C") do
+        expect(ENV["LC_ALL"]).to eq("C")
+      end
+
+      expect(ENV["LC_ALL"]).to eq("en_US.UTF-8")
+    end
+
+    specify "#run_as_not_developer" do
+      ENV["HOMEBREW_DEVELOPER"] = "foo"
+
+      run_as_not_developer do
+        expect(ENV["HOMEBREW_DEVELOPER"]).to be nil
+      end
+
+      expect(ENV["HOMEBREW_DEVELOPER"]).to eq("foo")
+    end
+
+    describe "#which" do
+      let(:cmd) { dir/"foo" }
+
+      before(:each) do
+        FileUtils.touch cmd
+      end
+
+      after(:each) do
+        dir.rmtree
+      end
+
+      it "returns the first executable that is found" do
+        cmd.chmod 0744
+        expect(which(File.basename(cmd), File.dirname(cmd))).to eq(cmd)
+      end
+
+      it "skips non-executables" do
+        expect(which(File.basename(cmd), File.dirname(cmd))).to be nil
+      end
+
+      it "skips malformed path and doesn't fail" do
+        # 'which' should not fail if a path is malformed
+        # see https://github.com/Homebrew/legacy-homebrew/issues/32789 for an example
+        cmd.chmod 0744
+
+        # ~~ will fail because ~foo resolves to foo's home and there is no '~' user
+        path = ["~~", File.dirname(cmd)].join(File::PATH_SEPARATOR)
+        expect(which(File.basename(cmd), path)).to eq(cmd)
+      end
+    end
+
+    describe "#which_all" do
+      let(:cmd1) { dir/"foo" }
+      let(:cmd2) { dir/"bar/foo" }
+      let(:cmd3) { dir/"bar/baz/foo" }
+
+      before(:each) do
+        (dir/"bar/baz").mkpath
+
+        FileUtils.touch cmd2
+
+        [cmd1, cmd3].each do |cmd|
+          FileUtils.touch cmd
+          cmd.chmod 0744
+        end
+      end
+
+      after(:each) do
+        dir.rmtree
+      end
+
+      it "returns an array of all executables that are found" do
+        path = [
+          "#{dir}/bar/baz",
+          "#{dir}/baz:#{dir}",
+          "~baduserpath",
+        ].join(File::PATH_SEPARATOR)
+        expect(which_all("foo", path)).to eq([cmd3, cmd1])
+      end
+    end
+
+    specify "#which_editor" do
+      ENV["HOMEBREW_EDITOR"] = "vemate"
+      expect(which_editor).to eq("vemate")
+    end
+
+    specify "#gzip" do
+      Dir.mktmpdir do |path|
+        path = Pathname.new(path)
+        somefile = path/"somefile"
+        FileUtils.touch somefile
+        expect(gzip(somefile)[0].to_s).to eq("#{somefile}.gz")
+        expect(Pathname.new("#{somefile}.gz")).to exist
+      end
+    end
+
+    specify "#capture_stderr" do
+      expect(
+        capture_stderr do
+          $stderr.print "test"
+        end,
+      ).to eq("test")
+    end
+
+    describe Shell do
+      subject { described_class }
+
+      describe "#shell_profile" do
+        before(:each) do
+          ENV["SHELL"] = shell
+        end
+
+        context "default" do
```

Comment:
I think these contexts add unnecessary indirection, and these tests should just be single examples that set `ENV["SHELL"]`.